tests/test_app/test_install_plugins.py

```diff
@@ -0,0 +1,256 @@
+    check_call([sys.executable, "-m", "pip", "install", "."], cwd=temp_dir)
+
+    # Wait for the installation to complete
+    timeout = 60  # maximum time to wait in seconds
+    start_time = time.time()
+
+    start_time = time.time()
+    package_name = mod
+
+    while True:
+        if is_package_installed(package_name=package_name):
+            break
+        elif time.time() - start_time > timeout:
+            raise TimeoutError(
+                f"Package {package_name!r} not found after waiting for {timeout} seconds."
+            )
+        time.sleep(1)
+
+    # check we can import it
+    timeout = 60  # maximum time to wait in seconds
+    start_time = time.time()
+
+    while True:
+        with suppress(ImportError):
+            importlib.import_module(package_name)
+            break
+        if time.time() - start_time > timeout:
+            raise TimeoutError(
+                f"Package {package_name!r} not importable after waiting for {timeout} seconds."
+            )
+        time.sleep(1)
+
+    # check it's in the stevedore cache
+    timeout = 60  # maximum time to wait in seconds
+    start_time = time.time()
+
+    while True:
+        appcache = apps(force=True)
+        if any(ext.entry_point.value == f"{mod}:{cls.__name__}" for ext in appcache):
+            break
+        elif time.time() - start_time > timeout:
+            raise TimeoutError(
+                f"App {mod}.{cls.__name__} not available after waiting for {timeout} seconds."
+            )
+        time.sleep(1)
+
+
+def uninstall_app(cls, mod=None):
+    """
+    Uninstalls a temporary app created from a class by running pip uninstall -y in the temporary directory.
+    """
+    if not mod:
+        mod = f"mod_{cls.__name__}"
+    # Use subprocess to run pip uninstall -y in the temporary directory
+    package_name = mod.replace("_", "-")  # replace underscores with hyphens
+    check_call([sys.executable, "-m", "pip", "uninstall", "-y", package_name])
+    # force the app cache to reload
+    _ = available_apps(force=True)
+
+
+@contextmanager
+def temp_app(cls, module_name: str = None):
+    """
+    A context manager that creates a temporary app from a class, installs it,
+    and then uninstalls it after testing.
+
+    Parameters
+    ----------
+    cls : object
+        The class from which to create the app.
+    module_name : str, optional
+        The name of the module in which the app is defined. If None, a name
+        is generated based on the class name.
+
+    Yields
+    ------
+    None
+
+    Raises
+    ------
+    pytest.fail
+        If a TimeoutError occurs during app installation.
+
+    Notes
+    -----
+    The app is installed in a temporary directory, which is deleted after testing.
+
+    Usage
+    -----
+    with temp_app(cls):
+        myapp = get_app(cls.__name__)
+        # test myapp
+    """
+    if module_name is None:
+        module_name = f"mod_{cls.__name__}"
+    temp_dir = mkdtemp()

```

Comment:
is there a reason you can't use pytest temp_path fixture here so you don't manually create `mkdtemp()`? This approach can be flaky if the default temp directory is very small (e.g. on a VM with a secondary and much larger mount). Use `pytest` so any changes to settings can be done in one place

Reply:
That was just laziness on my part to not revert to pytest temp_path.

I was seeing what appeared to be a race condition with stevedore's cache not being updated after the second installation of a temp app.  So I was playing with when I removed the temp directory in case that was the problem.  Turns out the problem was not a race condition, but a change in stevedore to use importlib instead of pkg_resources, and an implicit assumption in the python "use pattern" that the interpreter will be reloaded between package installation.  Which leaves this as a live problem for multiple dynamic loading of apps for testing purposes.

Will change to use pytest temp_path.  Looking into strategies for hiving off dynamic loading of apps that can't be addressed with a mocked ExtensionManager (ie: collision with existing tests) to a special test suite (like we do with "slow" tests).